core/io/file_access_memory.cpp

```diff
@@ -133,8 +133,13 @@ uint64_t FileAccessMemory::get_buffer(uint8_t *p_dst, uint64_t p_length) const {
 		WARN_PRINT("Reading less data than requested");
 	}
 
-	memcpy(p_dst, &data[pos], read);
-	pos += read;
+	if (read) {
+		// Don't call memcpy on a potential nullptr dest
+		// even when read is 0 as this triggers undefined
+		// behavior and may cause memory corruption.
+		memcpy(p_dst, &data[pos], read);
```

Comment:
That's what I meant by my previous comment, checks are logically fine, but comment is not clear. Changing the check to `if (src && dst)` will work the same, but it unambiguously shows what's checked, no comments needed.

> Ah, indeed, I missed p_length part. But still, something like if (ptr && p_src) might be cleaner than checking size.

Reply:
It this case it probably worth doing for readability, but also the whole size calculation can be moved into the same if block:

```c++
	uint64_t read  = 0;
	if (p_dst && data) {
		uint64_t left = length - pos;
		read = MIN(p_length, left);

		if (read < p_length) {
			WARN_PRINT("Reading less data than requested");
		}

		memcpy(p_dst, &data[pos], read);
		pos += read;
	}
	return read;
```

Reply:
That being said I think it's also fine to keep it simple like done in this PR (avoid regressions), and refactor the code for readability in a follow-up.

Reply:
> This seems to be an expectation in the codebase indeed, which I believe mirrors memcpy?

That is what I codebase implies, but calling memcpy with null parameters is undefined behavior. It makes gcc remove subsequent nullptr checks because calling memcpy on a pointer proves it is not nullptr, because that would be UB.